File: backend/user-service/src/main/java/com/agrima/user/service/AdminService.java
package com.agrima.user.service;

import com.agrima.user.dto.*;
import com.agrima.user.model.User;
import com.agrima.user.model.ActionModeration;
import com.agrima.user.repository.UserRepository;
import com.agrima.user.repository.ActionModerationRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class AdminService {

    private final UserRepository userRepository;
    private final ActionModerationRepository actionModerationRepository;
    private final NotificationService notificationService;
    private final AdminMapper adminMapper = AdminMapper.INSTANCE;

    // ========== GESTION DES PROFILS ==========

    public Page<ProfilAdminDTO> obtenirProfilsFiltres(String role, String statut, Pageable pageable) {
        log.info("Récupération des profils - role: {}, statut: {}", role, statut);
        
        Page<User> profils;
        if (role != null && statut != null) {
            profils = userRepository.findByRoleAndStatut(role, statut, pageable);
        } else if (role != null) {
            profils = userRepository.findByRole(role, pageable);
        } else if (statut != null) {
            profils = userRepository.findByStatut(statut, pageable);
        } else {
            profils = userRepository.findAll(pageable);
        }
        
        return profils.map(adminMapper::toProfilAdminDTO);
    }

    public ProfilAdminDTO obtenirProfil(Long id) {
        log.info("Récupération du profil: {}", id);
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));
        return adminMapper.toProfilAdminDTO(user);
    }

    public List<ProfilAdminDTO> obtenirProfilsUrgents() {
        log.info("Récupération des profils urgents");
        List<User> profils = userRepository.findByNbSignalementsGreaterThanOrStatut(3, "Signalé");
        return profils.stream()
            .map(adminMapper::toProfilAdminDTO)
            .collect(Collectors.toList());
    }

    public ProfilAdminDTO mettreAJourProfil(Long id, UpdateProfilRequest request) {
        log.info("Mise à jour du profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        if (request.getTelephone() != null) {
            user.setTelephone(request.getTelephone());
        }
        if (request.getLocalisation() != null) {
            user.setLocalisation(request.getLocalisation());
        }
        if (request.getStatut() != null) {
            user.setStatut(request.getStatut());
        }
        if (request.getScoreConfiance() != null) {
            user.setScoreConfiance(request.getScoreConfiance());
        }

        User updated = userRepository.save(user);
        
        log.info("Profil mis à jour avec succès: {}", id);
        return adminMapper.toProfilAdminDTO(updated);
    }

    public void supprimerProfil(Long id) {
        log.info("Suppression du profil: {}", id);
        userRepository.deleteById(id);
    }

    // ========== ACTIONS DE MODERATION ==========

    public ProfilAdminDTO validerProfil(Long id, String raison) {
        log.info("Validation du profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        user.setStatut("Validé");
        User updated = userRepository.save(user);

        ActionModeration action = ActionModeration.builder()
            .user(user)
            .type("VALIDATION")
            .description("Profil validé par l'administrateur")
            .raison(raison)
            .dateAction(LocalDateTime.now())
            .statut("EFFECTUEE")
            .build();
        actionModerationRepository.save(action);

        notificationService.notifierValidation(user);

        return adminMapper.toProfilAdminDTO(updated);
    }

    public ProfilAdminDTO suspendreTemporairement(Long id, Integer duree, String raison) {
        log.info("Suspension temporaire du profil: {} pour {} jours", id, duree);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        user.setStatut("Suspendu");
        user.setDateSuspension(LocalDateTime.now());
        user.setDureeSuspension(duree);
        User updated = userRepository.save(user);

        ActionModeration action = ActionModeration.builder()
            .user(user)
            .type("SUSPENSION")
            .description("Profil suspendu temporairement pour " + duree + " jours")
            .raison(raison)
            .dateAction(LocalDateTime.now())
            .statut("EFFECTUEE")
            .build();
        actionModerationRepository.save(action);

        notificationService.notifierSuspension(user, duree, raison);

        return adminMapper.toProfilAdminDTO(updated);
    }

    public ProfilAdminDTO reactiverProfil(Long id) {
        log.info("Réactivation du profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        user.setStatut("Actif");
        user.setDateSuspension(null);
        user.setDureeSuspension(null);
        User updated = userRepository.save(user);

        ActionModeration action = ActionModeration.builder()
            .user(user)
            .type("REACTIVATION")
            .description("Profil réactivé")
            .dateAction(LocalDateTime.now())
            .statut("EFFECTUEE")
            .build();
        actionModerationRepository.save(action);

        notificationService.notifierReactivation(user);

        return adminMapper.toProfilAdminDTO(updated);
    }

    public ActionModerationDTO ajouterAvertissement(Long id, String raison, String details) {
        log.info("Ajout d'un avertissement au profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        ActionModeration action = ActionModeration.builder()
            .user(user)
            .type("AVERTISSEMENT")
            .description("Avertissement: " + details)
            .raison(raison)
            .dateAction(LocalDateTime.now())
            .statut("EFFECTUEE")
            .build();
        ActionModeration saved = actionModerationRepository.save(action);

        notificationService.notifierAvertissement(user, raison);

        return adminMapper.toActionModerationDTO(saved);
    }

    public ProfilAdminDTO bloquerProfil(Long id, String raison) {
        log.info("Blocage du profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        user.setStatut("Bloqué");
        user.setDateBlocage(LocalDateTime.now());
        User updated = userRepository.save(user);

        ActionModeration action = ActionModeration.builder()
            .user(user)
            .type("BLOCAGE")
            .description("Profil bloqué définitivement")
            .raison(raison)
            .dateAction(LocalDateTime.now())
            .statut("EFFECTUEE")
            .build();
        actionModerationRepository.save(action);

        notificationService.notifierBlocage(user, raison);

        return adminMapper.toProfilAdminDTO(updated);
    }

    public MessageDTO envoyerMessage(Long id, String sujet, String contenu) {
        log.info("Envoi d'un message au profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        ActionModeration action = ActionModeration.builder()
            .user(user)
            .type("CONTACT")
            .description("Message: " + sujet)
            .raison(contenu)
            .dateAction(LocalDateTime.now())
            .statut("EFFECTUEE")
            .build();
        actionModerationRepository.save(action);

        notificationService.envoyerMessage(user, sujet, contenu);

        return MessageDTO.builder()
            .profilId(id)
            .sujet(sujet)
            .contenu(contenu)
            .dateEnvoi(LocalDateTime.now())
            .lu(false)
            .build();
    }

    public List<ActionModerationDTO> obtenirHistoriqueActions(Long id) {
        log.info("Récupération de l'historique des actions du profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        List<ActionModeration> actions = actionModerationRepository.findByUser(user);
        return actions.stream()
            .map(adminMapper::toActionModerationDTO)
            .collect(Collectors.toList());
    }

    // ========== GESTION DES ROLES ==========

    public ProfilAdminDTO assignerRole(Long id, String role) {
        log.info("Assignation du rôle {} au profil: {}", role, id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        user.setRole(role);
        User updated = userRepository.save(user);

        log.info("Rôle assigné avec succès");
        return adminMapper.toProfilAdminDTO(updated);
    }

    public ProfilAdminDTO retirerRole(Long id, String role) {
        log.info("Retrait du rôle {} au profil: {}", role, id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        User updated = userRepository.save(user);
        
        log.info("Rôle retiré avec succès");
        return adminMapper.toProfilAdminDTO(updated);
    }

    // ========== ACTIONS EN ATTENTE ==========

    public List<ActionModerationDTO> obtenirActionsEnAttente() {
        log.info("Récupération des actions en attente");
        
        List<ActionModeration> actions = actionModerationRepository.findByStatut("EN_ATTENTE");
        return actions.stream()
            .map(adminMapper::toActionModerationDTO)
            .collect(Collectors.toList());
    }

    public ActionModerationDTO approuverAction(Long id) {
        log.info("Approbation de l'action: {}", id);
        
        ActionModeration action = actionModerationRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Action non trouvée: " + id));

        action.setStatut("EFFECTUEE");
        ActionModeration updated = actionModerationRepository.save(action);

        return adminMapper.toActionModerationDTO(updated);
    }

    public ActionModerationDTO rejeterAction(Long id, String raison) {
        log.info("Rejet de l'action: {}", id);
        
        ActionModeration action = actionModerationRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Action non trouvée: " + id));

        action.setStatut("REJETEE");
        action.setRaison(raison);
        ActionModeration updated = actionModerationRepository.save(action);

        return adminMapper.toActionModerationDTO(updated);
    }

    // ========== ANALYTICS ET RAPPORTS ==========

    public StatistiquesGlobalesDTO obtenirStatistiques() {
        log.info("Calcul des statistiques globales");
        
        Long total = userRepository.count();
        Long actifs = userRepository.countByStatut("Actif");
        Long suspendus = userRepository.countByStatut("Suspendu");
        Long enAttente = userRepository.countByStatut("En attente");
        Double scoreMoyen = userRepository.getScoreConfiantMoyen();
        Long signalements = userRepository.countSignalements();

        return StatistiquesGlobalesDTO.builder()
            .totalProfils(total.intValue())
            .actifs(actifs.intValue())
            .suspendus(suspendus.intValue())
            .enAttente(enAttente.intValue())
            .scoreConfiantMoyen(scoreMoyen != null ? scoreMoyen : 0.0)
            .signalementsTotaux(signalements.intValue())
            .build();
    }

    public RapportModerationDTO genererRapportModeration(LocalDate dateDebut, LocalDate dateFin) {
        log.info("Génération du rapport de modération de {} à {}", dateDebut, dateFin);
        
        Long totalActions = actionModerationRepository.countByDateActionBetween(
            dateDebut.atStartOfDay(),
            dateFin.atTime(23, 59, 59)
        );

        return RapportModerationDTO.builder()
            .dateDebut(dateDebut.toString())
            .dateFin(dateFin.toString())
            .actionsEffectuees(totalActions.intValue())
            .build();
    }

    public ProfilExportDTO exporterProfil(Long id) {
        log.info("Export du profil: {}", id);
        
        User user = userRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Profil non trouvé: " + id));

        List<ActionModeration> actions = actionModerationRepository.findByUser(user);

        return ProfilExportDTO.builder()
            .id(user.getId())
            .nom(user.getNom())
            .email(user.getEmail())
            .telephone(user.getTelephone())
            .role(user.getRole())
            .statut(user.getStatut())
            .dateInscription(user.getDateInscription())
            .historique(actions.stream()
                .map(adminMapper::toActionModerationDTO)
                .collect(Collectors.toList()))
            .build();
    }

}

File: frontend/web/src/app/services/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { environment } from '../../environments/environment';

export interface ProfilAdmin {
  id: number;
  nom: string;
  email: string;
  telephone: string;
  role: string;
  statut: string;
  localisation: string;
  dateInscription: string;
  derniereActivite: string;
  scoreConfiance: number;
  nbSignalements: number;
  note: number;
  portefeuille?: string;
  tags: string[];
  siret?: string;
  nomExploitation?: string;
  certifications?: string[];
  vehicule?: string;
  plaqueImmatriculation?: string;
  adresse?: string;
  codePostal?: string;
  ville?: string;
}

export interface ActionModeration {
  id: number;
  profilId: number;
  type: 'SUSPENSION' | 'AVERTISSEMENT' | 'VALIDATION' | 'CONTACT' | 'BLOCAGE';
  description: string;
  raison: string;
  dateAction: string;
  moderateurId?: number;
  statut: 'EFFECTUEE' | 'EN_ATTENTE' | 'REJETEE';
}

export interface RapportModeration {
  dateDebut: string;
  dateFin: string;
  totalProfils: number;
  actifs: number;
  suspendus: number;
  enAttente: number;
  scoreConfiantMoyen: number;
  signalementsTotaux: number;
  actionsEffectuees: number;
}

export interface Page<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

@Injectable({
  providedIn: 'root'
})
export class AdminService {
  private readonly baseUrl = `${environment.apiUrl}/api/users/admin`;

  private profilsSubject = new BehaviorSubject<ProfilAdmin[]>([]);
  public profils$ = this.profilsSubject.asObservable();

  private profilSelectionneSubject = new BehaviorSubject<ProfilAdmin | null>(null);
  public profilSelectionne$ = this.profilSelectionneSubject.asObservable();

  private actionsSubject = new BehaviorSubject<ActionModeration[]>([]);
  public actions$ = this.actionsSubject.asObservable();

  constructor(private http: HttpClient) {}

  // ========== GESTION DES PROFILS ==========

  chargerProfils(): void {
    this.obtenirProfilsFiltres().subscribe();
  }

  obtenirProfil(id: number): Observable<ProfilAdmin> {
    return this.http.get<ProfilAdmin>(`${this.baseUrl}/profils/${id}`);
  }

  obtenirProfilsFiltres(role?: string, statut?: string): Observable<ProfilAdmin[]> {
    let params = new HttpParams();
    if (role && role !== 'TOUS') {
      params = params.set('role', role);
    }
    if (statut && statut !== 'TOUS') {
      params = params.set('statut', statut);
    }

    return this.http.get<Page<ProfilAdmin>>(`${this.baseUrl}/profils`, { params }).pipe(
      map(page => page.content),
      tap(profils => this.profilsSubject.next(profils))
    );
  }

  obtenirProfilsUrgents(): Observable<ProfilAdmin[]> {
    return this.http.get<ProfilAdmin[]>(`${this.baseUrl}/profils/urgents`);
  }

  selectionnerProfil(profil: ProfilAdmin): void {
    this.profilSelectionneSubject.next(profil);
  }

  obtenirConsommateur(id: number): Observable<any> {
    return this.http.get<any>(`${environment.apiUrl}/api/users/${id}/consommateur`);
  }

  obtenirProducteur(id: number): Observable<any> {
    return this.http.get<any>(`${environment.apiUrl}/api/users/${id}/producteur`);
  }

  obtenirLivreur(id: number): Observable<any> {
    return this.http.get<any>(`${environment.apiUrl}/api/users/${id}/livreur`);
  }

  mettreAJourProfil(id: number, donnees: Partial<ProfilAdmin>): Observable<ProfilAdmin> {
    return this.http.put<ProfilAdmin>(`${this.baseUrl}/profils/${id}`, donnees).pipe(
      tap(() => this.chargerProfils())
    );
  }

  supprimerProfil(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/profils/${id}`).pipe(
      tap(() => this.chargerProfils())
    );
  }

  // ========== ACTIONS DE MODERATION ==========

  validerProfil(id: number, raison: string): Observable<ProfilAdmin> {
    return this.http.post<ProfilAdmin>(`${this.baseUrl}/profils/${id}/valider`, { raison }).pipe(
      tap(() => this.chargerProfils())
    );
  }

  suspendreTemporairement(id: number, duree: number, raison: string): Observable<ProfilAdmin> {
    return this.http.post<ProfilAdmin>(`${this.baseUrl}/profils/${id}/suspendre-temp`, { duree, raison }).pipe(
      tap(() => this.chargerProfils())
    );
  }

  reactiverProfil(id: number): Observable<ProfilAdmin> {
    return this.http.post<ProfilAdmin>(`${this.baseUrl}/profils/${id}/reactiver`, {}).pipe(
      tap(() => this.chargerProfils())
    );
  }

  ajouterAvertissement(id: number, raison: string, details: string): Observable<ActionModeration> {
    return this.http.post<ActionModeration>(`${this.baseUrl}/profils/${id}/avertir`, { raison, details });
  }

  bloquerProfil(id: number, raison: string): Observable<ProfilAdmin> {
    return this.http.post<ProfilAdmin>(`${this.baseUrl}/profils/${id}/bloquer`, { raison }).pipe(
      tap(() => this.chargerProfils())
    );
  }

  envoyerMessage(id: number, sujet: string, contenu: string): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/profils/${id}/message`, { sujet, contenu });
  }

  // ========== GESTION DES ROLES ==========

  assignerRole(userId: number, nouveauRole: string): Observable<ProfilAdmin> {
    return this.http.post<ProfilAdmin>(`${this.baseUrl}/utilisateurs/${userId}/role`, { role: nouveauRole }).pipe(
      tap(() => this.chargerProfils())
    );
  }

  retirerRole(userId: number, role: string): Observable<ProfilAdmin> {
    return this.http.delete<ProfilAdmin>(`${this.baseUrl}/utilisateurs/${userId}/role/${role}`).pipe(
      tap(() => this.chargerProfils())
    );
  }

  // ========== ANALYTICS ET RAPPORTS ==========

  genererRapportModeration(dateDebut: string, dateFin: string): Observable<RapportModeration> {
    let params = new HttpParams().set('dateDebut', dateDebut).set('dateFin', dateFin);
    return this.http.get<RapportModeration>(`${this.baseUrl}/rapports/moderation`, { params });
  }

  obtenirStatistiques(): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/statistiques`);
  }

  obtenirHistoriqueActions(profilId: number): Observable<ActionModeration[]> {
    return this.http.get<ActionModeration[]>(`${this.baseUrl}/profils/${profilId}/historique-actions`);
  }

  obtenirActionsEnAttente(): Observable<ActionModeration[]> {
    return this.http.get<ActionModeration[]>(`${this.baseUrl}/actions/en-attente`);
  }

  approuverAction(actionId: number): Observable<ActionModeration> {
    return this.http.post<ActionModeration>(`${this.baseUrl}/actions/${actionId}/approuver`, {});
  }

  rejeterAction(actionId: number, raison: string): Observable<ActionModeration> {
    return this.http.post<ActionModeration>(`${this.baseUrl}/actions/${actionId}/rejeter`, { raison });
  }

  // ========== UTILITAIRES ==========

  calculerNiveauRisque(profil: ProfilAdmin): 'Faible' | 'Modéré' | 'Élevé' {
    if (profil.scoreConfiance >= 85 && profil.nbSignalements === 0) return 'Faible';
    if (profil.scoreConfiance >= 60 && profil.nbSignalements <= 2) return 'Modéré';
    return 'Élevé';
  }

  extraireInitiales(nom: string): string {
    if (!nom) return '';
    return nom
      .split(' ')
      .slice(0, 2)
      .map((part) => part.charAt(0).toUpperCase())
      .join('');
  }

  formaterDate(date: string | Date): string {
    if (!date) return '-';
    const d = new Date(date);
    return d.toLocaleDateString('fr-FR', {
      year: 'numeric',
      month: 'long',
      day: 'numeric'
    });
  }
}
